### PR TITLE
Feature/homework 05

### PR DESCRIPTION
**Pull request by igoryushchenko**

### CODE REVIEW

File: homework-04/config/config.json [hunk 1]
```diff
@@ -0,0 +1,10 @@
+{
```
Comment (AlexeyBarkow) [inline homework-04/config/config.json:1]: I'd rather move all these constants to .env file and install https://www.npmjs.com/package/dotenv package

File: homework-04/routes/authentication.js [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+const secret = config.tokenSecret
+
+router.post('/local', (req, res) => {
+  if (req.parsedQuery && req.parsedQuery.email) {
+    const user = usersDb.find(user => user.email === req.parsedQuery.email)
+    if (user && user.password === req.parsedQuery.password) {
+      let data = {
+          user: user
+      }
+      let token = jwt.sign(data, secret, { expiresIn: 20 })
```
Comment (AlexeyBarkow) [inline homework-04/routes/authentication.js]: Please, move this value to constants

File: homework-04/routes/authentication.js [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+    done(null, user)
+  } else {
+    done(null, false, 'Invalid email or password')
+  }
+}))
+
+passport.use(new FacebookStrategy({
+  clientID: config.facebookAppId,
+  clientSecret: config.facebookAppSecret,
+  callbackURL: 'http://localhost:9000/api/auth/facebook/callback',
```
Comment (AlexeyBarkow) [inline homework-04/routes/authentication.js]: this should be paced to config too

File: homework-04/routes/authentication.js [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+    } else {
+      console.log('User not found in DB')
+    }
+  }
+}))
+
+passport.use(new TwitterStrategy({
+  consumerKey: config.twitterConsumerKey,
+  consumerSecret: config.twitterConsumerSecret,
+  callbackURL: 'http://localhost:9000/api/auth/twitter/callback',
```
Comment (AlexeyBarkow) [inline homework-04/routes/authentication.js]: the same

File: homework-04/routes/authentication.js [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+    } else {
+      console.log('User not found in DB')
+    }
+  }
+}))
+
+passport.use(new GoogleStrategy({
+  clientID: config.googleConsumerKey,
+  clientSecret: config.googleConsumerSecret,
+  callbackURL: 'http://localhost:9000/api/auth/google/callback',
```
Comment (AlexeyBarkow) [inline homework-04/routes/authentication.js]: the same

File: homework-04/routes/authentication.js [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+      console.log('User not found in DB')
+    }
+  }
+}))
+
+router.post('/passport-local', passport.authenticate('local', { session: false }), (req, res) => {
+  let data = {
+    user: req.user
+  }
+  let token = jwt.sign(data, secret, { expiresIn: 20 })
```
Comment (AlexeyBarkow) [inline homework-04/routes/authentication.js]: move to constants

File: homework-04/routes/authentication.js [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+})
+
+router.get('/facebook', passport.authenticate('facebook', { session: false }))
+
+router.get('/facebook/callback',
+  passport.authenticate('facebook', { failureRedirect: '/facebook', session: false }), (req, res) => {
+    let data = {
+      user: req.user
+    }
+    let token = jwt.sign(data, secret, { expiresIn: 20 })
```
Comment (AlexeyBarkow) [inline homework-04/routes/authentication.js]: the same

File: homework-04/routes/authentication.js [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+
+const router = express.Router()
+
+const GoogleStrategy = require('passport-google-oauth').OAuth2Strategy
+
+const usersDb = User.getUsersDB()
+const secret = config.tokenSecret
+
+router.post('/local', (req, res) => {
+  if (req.parsedQuery && req.parsedQuery.email) {
```
Comment (AlexeyBarkow) [inline homework-04/routes/authentication.js]: don't use query for `post` requests, especially for credentials

File: homework-04/routes/authentication.js [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+}))
+
+passport.use(new GoogleStrategy({
+  clientID: config.googleConsumerKey,
+  clientSecret: config.googleConsumerSecret,
+  callbackURL: 'http://localhost:9000/api/auth/google/callback',
+  session: false
+}, (accessToken, refreshToken, profile, done) => {
+  if (profile) {
+    const user = usersDb.find(user => user.id === profile.id)
```
Comment (AlexeyBarkow) [inline homework-04/routes/authentication.js]: Please, add user to the db if it doesn't exist
or at least respond to my request with 403 status

**Review by AlexeyBarkow — CHANGES_REQUESTED** [review]

File: homework-04/config/config.json (hunk 1 above)
Comment (igoryushchenko) [reply homework-04/config/config.json:1]: fixed

File: homework-04/routes/authentication.js (hunk 2 above)
Comment (igoryushchenko) [reply homework-04/routes/authentication.js]: fixed

File: homework-04/routes/authentication.js (hunk 3 above)
Comment (igoryushchenko) [reply homework-04/routes/authentication.js]: fixed

File: homework-04/routes/authentication.js (hunk 4 above)
Comment (igoryushchenko) [reply homework-04/routes/authentication.js]: fixed

File: homework-04/routes/authentication.js (hunk 5 above)
Comment (igoryushchenko) [reply homework-04/routes/authentication.js]: fixed

File: homework-04/routes/authentication.js (hunk 6 above)
Comment (igoryushchenko) [reply homework-04/routes/authentication.js]: fixed

File: homework-04/routes/authentication.js (hunk 7 above)
Comment (igoryushchenko) [reply homework-04/routes/authentication.js]: fixed

File: homework-04/routes/authentication.js (hunk 8 above)
Comment (igoryushchenko) [reply homework-04/routes/authentication.js]: fixed

File: homework-04/routes/authentication.js (hunk 9 above)
Comment (igoryushchenko) [reply homework-04/routes/authentication.js]: fixed

File: homework-04/middlewares/verifyJWT.js [hunk 10]
```diff
@@ -0,0 +1,26 @@
+import jwt from 'jsonwebtoken'
+import dotenv from 'dotenv'
+
+dotenv.config()
```
Comment (AlexeyBarkow) [inline homework-04/middlewares/verifyJWT.js]: I believe tihs should be placed at very index.js file

File: homework-04/config/config.json (hunk 1 above)
Comment (AlexeyBarkow) [reply homework-04/config/config.json:1]: Please, create `.env.example` file. I can see .env is gitignored

**Review by AlexeyBarkow — CHANGES_REQUESTED** [review]

File: homework-04/config/config.json (hunk 1 above)
Comment (igoryushchenko) [reply homework-04/config/config.json:1]: resolved.

File: homework-04/middlewares/verifyJWT.js (hunk 10 above)
Comment (igoryushchenko) [reply homework-04/middlewares/verifyJWT.js]: resolved

**Pull request merged**